src/Kupmios/KupmiosM/Ogmios.purs

```diff
@@ -159,11 +160,30 @@ ogmiosPostRequest
   :: Aeson -- ^ JSON-RPC request body
   -> KupmiosM (Either Affjax.Error (Affjax.Response String))
 ogmiosPostRequest body = do
-  config <- asks (_.ogmiosConfig <<< _.config)
+  { config: { ogmios }, ogmiosRequestSemaphore } <- ask
   logTrace' $ "sending ogmios HTTP request: " <> show body
-  s <- liftAff $ ogmiosPostRequestAff config body
-  logTrace' $ "response: " <> (show $ hush $ s)
-  pure s
+  let request = ogmiosPostRequestAff ogmios.serverConfig body
+  resp <-
+    liftAff case ogmiosRequestSemaphore of
+      Just sem ->
+        let
+          acquireSem =
+            case ogmios.requestSemaphoreCooldown of
+              Nothing ->
+                BoundedQueue.read sem
+              Just cd ->
+                BoundedQueue.isEmpty sem >>=
+                  case _ of
+                    false ->
+                      BoundedQueue.read sem
+                    true ->
+                      delay cd *> BoundedQueue.read sem
+        in
+          bracket acquireSem (const (BoundedQueue.write sem unit)) $ const request
+      Nothing ->
+        request
```

Comment:
I would abstract this away into a `rateLimited` function which is not coupled to Ogmios, maybe even put it in another module.

Reply:
Extracted the semaphore logic into the `rateLimited` function as suggested.

Reply:
Also, given the simplicity and effectiveness of the current solution, I don't see any reason to introduce additional dependencies.